### ISSUE TITLE
Maybe I found why the "Hybrid" weight is stronger than original weight

### ISSUE THREAD

**Issue by pangafu** [leela-zero/leela-zero#908]
As we mentioned in #814  #867 , I download 2 weights:
72ea669da5f491458cb5dfc44a80f0e760a9df71a8d0690b185822de49e00cb3.txt
af3c6e330b932c97b0e517ac7db544d37fe1d18184f101d4b0340ff02f51ce88.txt

And I made a "Hybrid" weight af3-72e_1-1.txt, it average parameters of upper 2 weight.

Then I open a new game, make some move, and see the heatmap of 3 weights in same situation
The test command is (playout 1):  leelaz.exe -g --noponder -t1 -p 1 -w weight.txt
![broad](https://user-images.githubusercontent.com/10149168/36576109-89bd51cc-1889-11e8-8d11-d3846c33ea8f.png)

I found the hybrid weight output is very closed to 2 original weights output's average, the result can see bellow.
![hybrid](https://user-images.githubusercontent.com/10149168/36576110-8a1e88c0-1889-11e8-9ac2-91d078078e7b.png)

So in my opinion, because the network is linear( or close to linear in small distance), we average the network's parameter, just equal to average several network's outputs. So "Hybrid" weight maybe equal to assemble and average many network's output, this method had been widely used to make predict more accurate.

**Comment by mylyu** [comment]
Really? I think our networks are not linear, as the activation functions are not.

**Comment by pangafu** [comment]
But in my test, the network maybe linear, or maybe linear in many situation?

**Comment by Splee99** [comment]
So after averaging the search tree may get wider. Given the same think time, it is up to the algorithm how the tree is searched and may not be more accurate in finding the best move. Remember in many cases we want to trim the tree to make search more efficient.

**Comment by Eddh** [comment]
On the other hand the agz paper talked about approximation errors that are mitigated by the averaging of leaf nodes evaluations. Maybe this also correct these approximation errors.

**Comment by Marcin1960** [comment]
Averaging a few noisy signals, makes signal MUCH sharper.

**Comment by Eddh** [comment]
In that case, this hybrid approach might not be a simple equivalent to lower learning rate.

**Comment by Dorus** [comment]
So how many games did this new net play vs the current best? How many did it win?

**Comment by wctgit** [comment]
Even if neural nets are non-linear in general, since they are differentiable (I'm guessing, since gradient descent is presumably being used), then nearby variations may perform as close-to-linear approximate combinations.

**Comment by marcocalignano** [comment]
Most functions are differentiable and very few are linear. And I do not know what "close-to-linear" is?

**Comment by wctgit** [comment]
Meaning that a tangent hyperplane to the curve closely approximates the curve near the point of tangency. The point being that small changes in the input point give small changes in the function evaluation. So two nearby points will have a near linear relationship between their function values along the line that joins them. But it will only be approximately linear, i.e. 'close-to-linear'.

**Comment by marcocalignano** [comment]
But if you want to approximate a function with an hyper-plane you need to guarantee that all the function inputs, at all time, come from the interval where the approximations gives you still an acceptable error.
I am sure the network is not linear, probably not derivable (remember there are composition of convolutions involved), and even if you could approximate with some linear function the intervall would be so small that almost all our input is out of it.

**Comment by wctgit** [comment]
That's what I originally thought, i.e. that taking weighted averages of networks wouldn't likely produce usable results. But recent empirical evidence shows that my original assessment was wrong. I'm just giving a probable explanation for why it _does_ actually work.

**Comment by pangafu** [comment]
@Dorus please see 5x64 Hybrid test in #867 , the 6x192 test still in process.

**Comment by pangafu** [comment]
@Eddh yes, I also think hybrid is not the learning rate problem.

**Comment by pangafu** [comment]
@marcocalignano I also think the network is none-linear, but the result show that maybe network is close-to-linear in small distance, maybe relu activation functions or other functions can be considered as linear in small distance.

**Comment by jkabrg** [comment]
@marcocalignano Convolutions are linear. The whole point of differentiability is that it allows for approximation by linear functions.

It's false that most functions are differentiable; the opposite is true.

ReLU is non-differentiable at some point, but it's still [sub-differentiable](https://en.wikipedia.org/wiki/Subderivative) at that point.

**Comment by RavnaBergsndot** [comment]
I think the fundamental reason is that the network already evolved to a state that no big structural change happens, and most of the filters' functionalities are already fixed. We pick two filters at the same layer and index from two different networks, average them, and it works, because they almost do the same thing.

**Comment by emc2314** [comment]
IMHO, hybrid is the learning rate problem if and only if you hybrid **ALL** trained networks in one iteration. Since you've deprecated the weaker networks when hybrid, there must be some improvement.

**Comment by Eddh** [comment]
https://arxiv.org/pdf/1605.06431.pdf This article seems very interesting to understand better the nets used in LZ and Alphago.

**Comment by herazul** [comment]
Just out of curiosity, did anyone tried to hybrid recent network to see how it does ?

**Comment by pangafu** [comment]
@herazul We kept try to hybrid network when new one is out, and it always show stronger than parent network.

**Comment by pcengine** [comment]
@pangafu Did you try to hybrid recent 192*10 networks? If yes, could you please offer some hybrid networks here? So people like me without too much computer knowledge, can download it and give it a try?

**Comment by pangafu** [comment]
@pcengine you can download the hybrid file here, it hybrid 3 10*192f with 1:0.8:0.8 mix parameter
and it show much stronger than original one.
https://pan.baidu.com/s/1B4YExfL2WdMM-PYM5C2cIg

**Comment by pcengine** [comment]
@pangafu  I tried your link, but it says error-404 (it doesn't exist). Could you please double check it?

**Comment by pangafu** [comment]
@pcengine maybe you block by chinease GFW.... I can open the link.
Or you can use other browser to try.

**Comment by pcengine** [comment]
@pangafu 
This is what I got when I tried your link (I'd bet you can read the info in the attached image file). It seems I can connect to baidu cloud, but not to the specific file you offered.  I also tried to use several different VPNs, and all returned the same error-404

![1](https://user-images.githubusercontent.com/12437584/37402356-2f05d816-27c6-11e8-822a-5e1053864117.png)

**Comment by pheasant75** [comment]
That link works for me.  
I am downloading it (very slow) and then I can share with google drive.

**Comment by pangafu** [comment]
@pheasant75 so please share to @pcengine

**Comment by pheasant75** [comment]
google drive link for w21.zip:
https://drive.google.com/open?id=1YKN4jSlFdyBaONIYBDPHsacWhB9xtd6Z

**Comment by ThorAvaTahr** [comment]
When you average rock and paper, do you win against scissors?

**Comment by MartinDevelopment** [comment]
@ThorAvaTahr It should since it takes all the properties and combines them together and averages it.

**Comment by pheasant75** [comment]
@ThorAvaTahr Model ensembling is a common and popular method to improve accuracy in machine learning.  See some explanations and variants of ensembling strategies.
https://mlwave.com/kaggle-ensembling-guide/

Currently what's done here are average and some manual test of different weights.  Technically, you can train another NN to come up with the best ensembling or even dynamically choose result from different network during a game.  Of course, that is only worth trying after we reach our training limit (when everything is stalled).

**Comment by ThorAvaTahr** [comment]
@pheasant75  Thanks for the link.  Well it is certainly a well known method and applied widely, e.g. in weather modeling. It is off course especially relevant in cases where models departing from different principles are involved. In this case the ensemble should average out structural biases of each of the individual methods. 

In this case I wonder. I seems that the ensemble of individual networks each trained in the same way can only mean that we haven't reached the optimum yet. My reasoning is there is a continuous path between the weights of each of the individual networks (after all they have a common ancestor from which they are trained). That must mean that there is a continuous path to the ensemble network and this path can always be found by gradient descent (given small enough learning rates and sufficient steps).

**Comment by ThorAvaTahr** [comment]
In principle it would be possible that each of the individual networks are sitting in twisting and curving landscape each in their own minimum separated from each other by local maxima, however, this is practically extremely unlikely in the very high dimensional space of the weights.

**Comment by Eddh** [comment]
@ThorAvaTahr https://arxiv.org/abs/1803.05407 This paper explains a lot about all that.

**Comment by sethtroisi** [comment]
closing, no active discussion for >6months with subsequent issues and PRs
